### PR TITLE
feat(plugin-conversation): add room avatar support

### LINKED ISSUE #203: Add room avatar support

### PR DESCRIPTION
**Pull request by ianwremmel**
fixes #203

### CODE REVIEW

File: packages/plugin-conversation/src/conversation.js [hunk 1]
```diff
@@ -453,6 +453,24 @@ const Conversation = SparkPlugin.extend({
...
+    return this._inferConversationUrl(conversation)
+      .then(() => {
+        const activity = ShareActivity.create(conversation, null, this.spark);
+        activity.add(avatar);
+
+        return this.prepare(activity, {
+          target: this.prepareConversation(conversation)
+        });
+      })
+      .then((a) => {
```
Comment (aimeex3) [inline packages/plugin-conversation/src/conversation.js]: do we need to check that the file is <1MB? or is server supposed to do that? and makes sure that files.items.length is 1

File: packages/plugin-conversation/src/conversation.js [hunk 2]
```diff
@@ -453,6 +453,24 @@ const Conversation = SparkPlugin.extend({
     return ShareActivity.create(conversation, null, this.spark);
   },
 
+  assign(conversation, avatar) {
+    return this._inferConversationUrl(conversation)
+      .then(() => {
+        const activity = ShareActivity.create(conversation, null, this.spark);
+        activity.add(avatar);
```
Comment (aimeex3) [inline packages/plugin-conversation/src/conversation.js]: this will upload the thumbnail as well right? but not sure if we need it

Comment (aimeex3) [inline packages/plugin-conversation/src/conversation.js]: this was the spec (similar to share activity but some props taken out) https://sqbu-github.cisco.com/WebExSquared/cloud-apps/wiki/Custom-Room-Avatar---Design-Spec

File: packages/plugin-conversation/src/conversation.js [hunk 3]
```diff
@@ -514,6 +542,21 @@ const Conversation = SparkPlugin.extend({
       .then((res) => res.body);
   },
 
+  unassign(conversation, activity) {
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/conversation.js]: This function needs documentation.

File: packages/plugin-conversation/src/process-image.browser.js [hunk 4]
```diff
@@ -52,7 +52,7 @@ export function computeDimensions({width, height}, maxWidth, maxHeight) {
  * @param {Number} thumbnailMaxHeight
  * @returns {Promise<Array>} Buffer, Dimensions, thumbnailDimensions
  */
-export default function processImage(file, thumbnailMaxWidth, thumbnailMaxHeight) {
+export default function processImage({file, thumbnailMaxWidth, thumbnailMaxHeight, enableThumbnails}) {
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/process-image.browser.js]: Let's add documentation for this new param and the update to using a single object argument

File: packages/plugin-conversation/src/conversation.js [hunk 5]
```diff
@@ -460,6 +460,34 @@ const Conversation = SparkPlugin.extend({
...
+   * @returns {Promise<Activity>}
+   */
+  assign(conversation, avatar) {
+    if ((avatar.size || avatar.length) > 1024 * 1024) {
+      return Promise.reject(new Error(`Room avatars must be less than 1MB`));
+    }
+    return this._inferConversationUrl(conversation)
+      .then(() => {
+        const activity = ShareActivity.create(conversation, null, this.spark);
+        activity.enableThumbnails = false;
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/conversation.js]: You could probably pass this in when you make the share, no?

```js
const activity = ShareActivity.create(conversation, {enableThumbnails: false}, this.spark);
```

File: packages/plugin-conversation/src/encryption-transforms.js [hunk 6]
```diff
@@ -167,6 +167,17 @@ export const transforms = toArray(`outbound`, {
...
+  encryptAssignActivity: {
+    direction: `outbound`,
+    fn(ctx, key, activity) {
+      return ctx.transform(`encryptVerbActivity`, key, activity)
+        .then(() => {
+          key = key || activity[KEY];
+          activity.encryptionKeyUrl = key.uri || key;
+        });
+    }
+  },
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/encryption-transforms.js]: Some of these are starting to be pretty similar... Maybe an alias? :)

File: packages/plugin-conversation/src/process-image.js [hunk 7]
```diff
@@ -15,7 +15,7 @@ import {pick} from 'lodash';
  * @param {Logger} logger
  * @returns {Promise<Array>} Buffer, Dimensions, thumbnailDimensions
  */
-export default function processImage(file, thumbnailMaxWidth, thumbnailMaxHeight, logger) {
+export default function processImage({file, thumbnailMaxWidth, thumbnailMaxHeight, enableThumbnails, logger}) {
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/process-image.js]: Same docs update required here.

File: packages/plugin-conversation/test/integration/spec/verbs.js [hunk 8]
```diff
@@ -94,8 +95,45 @@ describe(`plugin-conversation`, function() {
...
+
+          assert.property(c.avatar, `files`);
+          assert.property(c.avatar.files, `items`);
+          assert.lengthOf(c.avatar.files.items, 1);
+          assert.property(c.avatar.files.items[0], `fileSize`);
+          assert.property(c.avatar.files.items[0], `mimeType`);
+          assert.property(c.avatar.files.items[0], `objectType`);
+          assert.property(c.avatar.files.items[0], `scr`);
+          assert.property(c.avatar.files.items[0], `url`);
+          assert.equal(c.avatar.objectType, `content`);
```
Comment (Josh-Dykstra) [inline packages/plugin-conversation/test/integration/spec/verbs.js]: I would say we could make this a custom asserter, but I don't anticipate us doing this much.

File: packages/plugin-conversation/src/conversation.js (hunk 5 above)
Comment (ianwremmel) [inline packages/plugin-conversation/src/conversation.js]: the middle param on share activity is the `object`, not an options object, so i'm not sure it makes sense to add `enableThumbnails` there

Comment (Josh-Dykstra) [inline packages/plugin-conversation/src/conversation.js]: Now that you say that, yeah it would work but it probably makes things confusing.

**Pull request merged**